rust/src/delta.rs

```diff
@@ -412,6 +415,37 @@ impl DeltaTable {
             }
         }
 
+        self.apply_logs_after_current_version().await?;
+
+        Ok(())
+    }
+
+    pub async fn update(&mut self) -> Result<(), DeltaTableError> {
+        match self.get_last_checkpoint().await {
+            Ok(last_check_point) => {
+                let should_restore = self.last_check_point.is_none() || 
+                    self.last_check_point.unwrap().version != last_check_point.version;
+
+                if should_restore {
+                    self.last_check_point = Some(last_check_point);
+                    self.restore_checkpoint(last_check_point).await?;
+                    self.version = last_check_point.version + 1;
+                }
```

Comment:
If we derive `Eq` and `PartialEq` trait for `CheckPoint`, then we can rewrite the this block of code without unwrap as:

```suggestion
               if self.last_check_point.is_none()
                   || self.last_check_point == Some(last_check_point)
               {
                    self.last_check_point = Some(last_check_point);
                    self.restore_checkpoint(last_check_point).await?;
                    self.version = last_check_point.version + 1;
                }
```

Reply:
Hmmm. This is a nice idea, but do you think the semantics are okay? Would all callers be okay by defining `version` as the metric of equality even when compared instances may have different `size` and `parts`? Ideally, it would not be possible to get into such a state I suppose.

Reply:
I have a local commit that uses `version` as equality metric if you think the semantics are okay. I can push if we are on the same page and agree on semantics. As it stands, I do not see any usages that would corrupt the checkpoint in a way that invalidates the version equality check.

Contents are:

```
impl PartialEq for CheckPoint {
    fn eq(&self, other: &Self) -> bool {
        self.version == other.version
    }
}

impl Eq for CheckPoint {}
```

Reply:
yeah, i think this is a reasonable performance optimization, i can't think of a case why size and parts would be different for the same version other than bugs.

Reply:
Nice. pushed code change.